csharp/extractor/Semmle.Extraction.CSharp/Entities/Types/Type.cs

```diff
@@ -347,6 +355,211 @@ public override bool Equals(object? obj)
+
+            private GenericsRecursionGraph(INamedTypeSymbol startSymbol)
+            {
+                this.startSymbol = startSymbol;
+
+                ComputeInstantiationClosure();
+                ComputeGraphEdges();
+            }
+
+            private void ComputeGraphEdges()
+            {
+                foreach (var reference in instantiationClosure)
+                {
+                    var definition = reference.OriginalDefinition;
+                    if (SymbolEqualityComparer.Default.Equals(reference, definition))
+                    {
+                        // It's a definition, so no edges
+                        continue;
+                    }
+
+                    for (var i = 0; i < reference.TypeArguments.Length; i++)
+                    {
+                        var target = definition.TypeParameters[i];
+                        if (reference.TypeArguments[i] is ITypeParameterSymbol source)
+                        {
+                            // non-expanding
+                            if (!edges.TryGetValue(source, out var targets))
+                            {
+                                targets = new List<(ITypeParameterSymbol, bool)>();
+                                edges.Add(source, targets);
+                            }
+                            targets.Add((target, false));
+                        }
+                        else if (reference.TypeArguments[i] is INamedTypeSymbol namedType)
+                        {
+                            // expanding
+                            var sources = GetAllNestedTypeParameters(namedType);
+                            foreach (var s in sources)
+                            {
+                                if (!edges.TryGetValue(s, out var targets))
+                                {
+                                    targets = new List<(ITypeParameterSymbol, bool)>();
+                                    edges.Add(s, targets);
+                                }
+                                targets.Add((target, true));
+                            }
+                        }
+                    }
+                }
+            }
+
+            private List<ITypeParameterSymbol> GetAllNestedTypeParameters(INamedTypeSymbol symbol)
+            {
+                var res = new List<ITypeParameterSymbol>();
+
+                foreach (var typeArgument in symbol.TypeArguments)
+                {
+                    if (typeArgument is ITypeParameterSymbol typeParameter)
+                    {
+                        res.Add(typeParameter);
+                    }
+                    else if (typeArgument is INamedTypeSymbol namedType)
+                    {
+                        res.AddRange(GetAllNestedTypeParameters(namedType));
+                    }
+                }
```

Comment:
```suggestion
                static void AddTypeParameters(INamedTypeSymbol symbol)
                {
                    foreach (var typeArgument in symbol.TypeArguments)
                    {
                        if (typeArgument is ITypeParameterSymbol typeParameter)
                        {
                            res.Add(typeParameter);
                        }
                        else if (typeArgument is INamedTypeSymbol namedType)
                        {
                            AddTypeParameters(namedType);
                        }
                    }
                }
                
                AddTypeParameters(symbol);
```

to avoid creating multiple `List` objects.